### PR TITLE
Disable Zstandard decompression on nvCOMP 2.4 and Pascal GPus

### PR DESCRIPTION
**Pull request by vuule**
## Description
Zstandard decompression in nvCOMP 2.4 can produce incorrect results on compute 6.0 GPUs. 
This PR disables the Zstandard decompression in this configuration.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.

### CODE REVIEW

**Comment by vuule** [comment]
@GregoryKimball should I update the table at https://docs.rapids.ai/api/cudf/nightly/user_guide/io.html in this PR? It should indicate ZSTD as stable with an asterisk for the disabled case.

**Review by mythrocks — APPROVED** [review]
LGTM!

File: cpp/src/io/comp/nvcomp_adapter.cpp [hunk 1]
```diff
@@ -152,6 +152,25 @@ size_t batched_decompress_temp_size(compression_type compression,
...
+
+#if NVCOMP_ZSTD_IS_DISABLED_ON_PASCAL
+  int device;
+  int cc_major;
+  int cc_minor;
+  CUDF_CUDA_TRY(cudaGetDevice(&device));
+  CUDF_CUDA_TRY(cudaDeviceGetAttribute(&cc_major, cudaDevAttrComputeCapabilityMajor, device));
+  CUDF_CUDA_TRY(cudaDeviceGetAttribute(&cc_minor, cudaDevAttrComputeCapabilityMinor, device));
+  CUDF_EXPECTS(cc_major != 6 or cc_minor != 1,
+               "Zstandard decompression is disabled on some Pascal GPUs");
```
Comment (galipremsagar) [inline cpp/src/io/comp/nvcomp_adapter.cpp:170]: Is the issue only on `6.1`? will `6.0` work fine?

Comment (vuule) [reply cpp/src/io/comp/nvcomp_adapter.cpp:170]: Yup. No repro on P100, which is compute capability 6.0.

Comment (galipremsagar) [reply cpp/src/io/comp/nvcomp_adapter.cpp:170]: I see, got it. Thanks 👍

File: cpp/src/io/comp/nvcomp_adapter.cpp [hunk 2]
```diff
@@ -152,6 +152,25 @@ size_t batched_decompress_temp_size(compression_type compression,
...
 }
 
+void check_is_zstd_enabled()
+{
+  CUDF_EXPECTS(NVCOMP_HAS_ZSTD_DECOMP, "nvCOMP 2.3 or newer is required for Zstandard compression");
+  CUDF_EXPECTS(NVCOMP_ZSTD_IS_STABLE or cudf::io::detail::nvcomp_integration::is_all_enabled(),
+               "Zstandard compression is experimental, you can enable it through "
+               "`LIBCUDF_NVCOMP_POLICY` environment variable.");
+
+#if NVCOMP_ZSTD_IS_DISABLED_ON_PASCAL
```
Comment (jbrennan333) [inline cpp/src/io/comp/nvcomp_adapter.cpp:162]: I don't see this being defined. Are we going to do that in a separate PR?

Comment (vuule) [reply cpp/src/io/comp/nvcomp_adapter.cpp:162]: ah, I accidentally removed the define before pushing the change. Will fix, thanks.

**Comment by mythrocks** [comment]
I think I reviewed this incorrectly. Having a closer look now.

File: cpp/src/io/comp/nvcomp_adapter.cpp [hunk 3]
```diff
@@ -166,7 +166,7 @@ void check_is_zstd_enabled()
   CUDF_CUDA_TRY(cudaGetDevice(&device));
   CUDF_CUDA_TRY(cudaDeviceGetAttribute(&cc_major, cudaDevAttrComputeCapabilityMajor, device));
   CUDF_CUDA_TRY(cudaDeviceGetAttribute(&cc_minor, cudaDevAttrComputeCapabilityMinor, device));
-  CUDF_EXPECTS(cc_major != 6 or cc_minor != 0,
+  CUDF_EXPECTS(cc_major != 6 or cc_minor != 1,
```
Comment (mythrocks) [inline cpp/src/io/comp/nvcomp_adapter.cpp:169]: Ah, right. Thanks for catching that.

File: cpp/src/io/comp/nvcomp_adapter.cpp (hunk 2 above)
Comment (vuule) [reply cpp/src/io/comp/nvcomp_adapter.cpp:170]: should be correct now

File: cpp/src/io/comp/nvcomp_adapter.cpp [hunk 4]
```diff
@@ -59,9 +59,17 @@
...
   (NVCOMP_MAJOR_VERSION == 2 and NVCOMP_MINOR_VERSION == 3 and NVCOMP_PATCH_VERSION >= 2)
-#define NVCOMP_ZSTD_IS_EXPERIMENTAL 0
+#define NVCOMP_ZSTD_IS_STABLE 1
 #else
-#define NVCOMP_ZSTD_IS_EXPERIMENTAL 1
+#define NVCOMP_ZSTD_IS_STABLE 0
+#endif
+
+// Issue https://github.com/NVIDIA/spark-rapids/issues/6614 impacts nvCOMP 2.4 ZSTD decompression on
+// compute 6.0
```
Comment (jbrennan333) [inline cpp/src/io/comp/nvcomp_adapter.cpp:68]: (nit) 6.1?

**Review by jbrennan333 — APPROVED** [review]
+1 this looks good.  Hopefully nvcomp-2.4.1 will fix this and we can update to allow it.

File: cpp/src/io/comp/nvcomp_adapter.cpp (hunk 4 above)
Comment (vuule) [inline cpp/src/io/comp/nvcomp_adapter.cpp:68]:
```suggestion
// compute 6.1
```

File: cpp/src/io/comp/nvcomp_adapter.cpp [hunk 5]
```diff
@@ -152,6 +160,25 @@ size_t batched_decompress_temp_size(compression_type compression,
   return temp_size;
 }
 
+void check_is_zstd_enabled()
```
Comment (ttnghia) [inline cpp/src/io/comp/nvcomp_adapter.cpp:163]: Probably this should be put into an anonymous namespace.

**Review by vyasr — APPROVED** [review]

File: cpp/src/io/comp/nvcomp_adapter.cpp (hunk 5 above)
Comment (vuule) [reply cpp/src/io/comp/nvcomp_adapter.cpp:163]: hm, there's many more functions here that could be in an anonymous namespace. If you don't mind, I would prefer to make this change in a future PR.

File: cpp/src/io/comp/nvcomp_adapter.cpp [hunk 6]
```diff
@@ -65,8 +65,8 @@
 #endif
 
 // Issue https://github.com/NVIDIA/spark-rapids/issues/6614 impacts nvCOMP 2.4 ZSTD decompression on
-// compute 6.1
-#if NVCOMP_MAJOR_VERSION > 2 or (NVCOMP_MAJOR_VERSION == 2 and NVCOMP_MINOR_VERSION >= 4)
+// compute 6.x
+#if NVCOMP_MAJOR_VERSION == 2 and NVCOMP_MINOR_VERSION == 4
```
Comment (jbrennan333) [inline cpp/src/io/comp/nvcomp_adapter.cpp:69]: Don't we need to check the patch version == 0 as well?
Otherwise this will disable it for 2.4.1 as well.

**Review by jbrennan333 — APPROVED** [review]
+1 looks good

File: cpp/src/io/comp/nvcomp_adapter.cpp (hunk 5 above)
Comment (ttnghia) [reply cpp/src/io/comp/nvcomp_adapter.cpp:163]: Sure.

**Review by ttnghia — APPROVED** [review]

**Review by jolorunyomi — APPROVED** [review]

**Comment by xingwenqiang** [comment]
Use this version to export zstd.parquet files. An exception occurs when reading with arrow while some page compressed size is larger than the size before. The similarity between these pages is that the last multiple elements are empty. Is this a usage problem?，Looking forward to your reply！

**Comment by xingwenqiang** [comment]
Use this version to export zstd.parquet files. An exception occurs when reading with arrow while some page compressed size is larger than the size before. The similarity between these pages is that the last multiple elements are empty. Is this a usage problem?，Looking forward to your reply！
